Sources/SourceKit/SourceKitServer.swift

```diff
@@ -365,11 +365,16 @@ extension SourceKitServer {
     workspace.documentManager.open(note.params)
 
     let textDocument = note.params.textDocument
-    workspace.buildSettings.registerForChangeNotifications(for: textDocument.url)
+    let url = textDocument.url
+    let language = textDocument.language
 
-    if let service = languageService(for: textDocument.url, textDocument.language, in: workspace) {
-      service.openDocument(note.params)
-    }
+    // Create the `service` before registering for change notifications in case the `BuildSystem`
+    // immediately callbacks to us (its delegate) with `FileBuildSettings` so we notify this
+    // `service` (which might have been newly created).
+    let service = languageService(for: url, language, in: workspace)
```

Comment:
We still have a synchronous `BuildSystem.toolchain(for: URL, ...)` method, which is used behind the covers in `languageService(for: URL, ...)`, but ultimately the build system will need to be able to change the toolchain as well.  Your change here reminded me of this.  The current code cannot handle the toolchain changing.

I'm starting to think we may need to go back to providing the toolchain along with the build settings, since once you get called back by the build system you should know both the compiler arguments and the preferred toolchain at the same time.  You can always open the file with the default toolchain in the meantime to get syntactic support if necessary, but that's a tradeoff against potentially wasting time if we would get the callback quickly.  E.g. if you can get the settings in a couple of milliseconds, maybe it's better to just wait, especially since starting a new toolchain language server is (relatively) expensive.

Reply:
Yeah I think it should be combined with the build settings or perhaps we can mirror similar functionality to toolchain changes? I'm not sure about delaying the opening of the file, this is tough - if the build system may take awhile, the user gets nothing even if the toolchains are the same.

Reply:
Outdated, this has been removed